### PR TITLE
refactor(assert): clearer typing

### PR DESCRIPTION
**Pull request by michaelfig**
Use Typescript 3.7 to provide the 'asserts XXX' type to the assert function.  This helps ensure that callers who assert don't get warned about a variable being falsy later in the control flow.

Also, add the `harden` global and SES typing of it, mainly as an example of how to do this elsewhere.

### CODE REVIEW

File: packages/assert/package.json [hunk 1]
```diff
@@ -49,6 +49,9 @@
     "env": {
       "es6": true
     },
+    "globals": {
```
Comment (katelynsills) [inline packages/assert/package.json:52]: So this is necessary to stop the `harden` errors?

File: packages/assert/src/assert.js [hunk 2]
```diff
@@ -1,9 +1,10 @@
 // Copyright (C) 2019 Agoric, under Apache License 2.0
 
-/* global harden */
-
 // @ts-check
 
+// eslint-disable-next-line spaced-comment
+/// <reference types="ses"/>
```
Comment (katelynsills) [inline packages/assert/src/assert.js]: What does this do?

**Review by katelynsills — COMMENTED** [review]
Is there anything in particular I should test out or look at to see the effects?

File: packages/assert/src/assert.js [hunk 3]
```diff
@@ -224,6 +226,7 @@ function assert(flag, optDetails = details`Check failed`) {
  * @param {*} actual The value we received
  * @param {*} expected What we wanted
  * @param {Details} [optDetails] The details to throw
+ * @returns {asserts actual is expected}
```
Comment (Chris-Hibbert) [inline packages/assert/src/assert.js]: Isn't the contents of returns' argument supposed to be type-like. is `asserts actual is expected` looks like the comment. What does assert return when it doesn't throw?

File: packages/assert/src/assert.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/assert/src/assert.js]: ... and is the triple slash intentional?

File: packages/assert/package.json (hunk 1 above)
Comment (michaelfig) [reply packages/assert/package.json:52]: Yes, to prevent `harden` from triggering eslint.

File: packages/assert/src/assert.js (hunk 2 above)
Comment (michaelfig) [reply packages/assert/src/assert.js]: CF: https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html#-reference-types-

This imports the types from the `ses` package, namely global `harden` with the correct, type-preserving return value.

File: packages/assert/src/assert.js (hunk 3 above)
Comment (michaelfig) [reply packages/assert/src/assert.js]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

It's legitimate syntax but doesn't actually work (doesn't check for equality, checks for membership in a type).  So, I'll remove this instance, but preserve the type of `assert` itself, which is similar but correct.

Comment (michaelfig) [reply packages/assert/src/assert.js]: I turned out adding as specific types as I could.

**Comment by michaelfig** [comment]
> Is there anything in particular I should test out or look at to see the effects?

I've added a build script to `package.json` that runs `tsc`, which in this case validates all the types against the assert source code.  Passing the CI tests is necessary, but the proof is in the pudding when you actually use the assert functions in your own code and try to turn on `@ts-check`.

**Comment by michaelfig** [comment]
Best example so far:

```js
// @ts-check
import { assert } from '@agoric/assert';
const val = 2;
assert.typeof(val, 'number');
console.log(val); // hover over 'val' in this line, you get 'number'
assert.typeof(val, 'string');
console.log(val); // hover over 'val' in this line, you get 'never'
```

**Pull request merged**